tools/src/main/scala/scala/scalanative/build/Build.scala

```diff
@@ -53,60 +58,87 @@ object Build {
    *  @return
    *    [[Config#artifactPath]], the path to the resulting native binary.
    */
-  def build(config: Config)(implicit scope: Scope): Path =
+  def build(config: Config)(implicit scope: Scope): Path = {
+    def makeKey(config: Config): Int = {
+      // skip the whole nativeLink process if the followings are unchanged since the previous build
+      // - build configuration
+      // - class paths' mtime
+      // - the output native binary ('s mtime)
+      // Since the NIR code is shipped in jars, we should be able to detect the changes in NIRs.
+      // One thing we miss is, we cannot detect changes in c libraries somewhere in `/usr/lib`.
+      val key = (
+        config.toString, // we use toString because `config` object contains Logger instance that changes for every invocation
+        config.classPath.map(getNewestMtime(_)),
+        getLastModifiedTimeMillis(config.artifactPath)
+      )
+      key.hashCode()
+    }
     config.logger.time("Total") {
       // called each time for clean or directory removal
       checkWorkdirExists(config)
 
       // validate Config
       val fconfig = Validator.validate(config)
+      val inputHash = makeKey(fconfig)
 
-      // find and link
-      val linked = {
-        val entries = ScalaNative.entries(fconfig)
-        val linked = ScalaNative.link(fconfig, entries)
-        ScalaNative.logLinked(fconfig, linked)
-        linked
-      }
+      if (Files.exists(fconfig.artifactPath) &&
+          previousBuildInputHash == inputHash) {
+        fconfig.logger.info(
+          "Build skipped: No changes detected in build configuration and class path contents since last build."
+        )
+        fconfig.artifactPath
+      } else {
```

Comment:
It might be useful to refactor the `else` branch into a separate method. That way we still would be able to access it while testing/benchmarking to ensure caching would not interfere with the results. We can make it public to allow other build tools to access it directly if they don't want to relay on our caching. So base on that we can rename current `build` to `buildCached`, and refactor the else part into the `build`

Reply:
You might consider using the following `buildImpl` code as it also solves the confusion of not using the correct config reference after filtering as well as having the ability to return with an `if` in the short code section. https://github.com/scala-native/scala-native/pull/3283

Reply:
I added `buildCache` endpoint, and left the `build` same, so build tools can choose which endpoint to use